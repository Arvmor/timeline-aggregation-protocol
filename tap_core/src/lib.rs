// Copyright 2023-, Semiotic AI, Inc.
// SPDX-License-Identifier: Apache-2.0


//! # Timeline Aggregation Protocol
//!
//! ## Overview
//!
//! The TAP (Timeline Aggregation Protocol) facilitates a series of payments from
//! a sender to a receiver, who aggregates these payments into a single payment.
//! This aggregate payment can then be verified on-chain by a payment verifier,
//! reducing the number of transactions and simplifying the payment process.
//!
//! ## Key Components
//!
//! - **Sender:** Initiates the payment.
//! - **Receiver:** Receives the payment.
//! - **Signers:** Multiple signers authorized by the sender to sign receipts.
//! - **State Channel:** A one-way channel opened by the sender with the receiver
//! for sending receipts.
//! - **Receipt:** A record of payment sent by the sender to the receiver.
//! - **ReceiptAggregateVoucher (RAV):** A signed message containing the aggregate
//! value of the receipts.
//! - **tap_aggregator:** An entity managed by the sender that signs RAV requests.
//! - **EscrowAccount:** An account created in the blockchain to hold funds for
//! the sender-receiver pair.
//!
//! ## Security Measures
//!
//! - The protocol uses asymmetric cryptography to sign and verify messages,
//! ensuring the integrity of receipts and RAVs.
//!
//! ## Process
//!
//! 1. **Opening a State Channel:** A state channel is opened via a blockchain
//! contract, creating an EscrowAccount for the sender-receiver pair.
//! 2. **Sending Receipts:** The sender sends receipts to the receiver through
//! the state channel.
//! 3. **Storing Receipts:** The receiver stores the receipts and tracks the
//! aggregate payment.
//! 4. **Creating a RAV Request:** A RAV request consists of a list of receipts
//! and, optionally, the previous RAV.
//! 5. **Signing the RAV:** The receiver sends the RAV request to the
//! tap_aggregator, which signs it into a RAV.
//! 6. **Tracking Aggregate Value:** The receiver tracks the aggregate value and
//! new receipts since the last RAV.
//! 7. **Requesting a New RAV:** The receiver sends new receipts and the last
//! RAV to the tap_aggregator for a new RAV.
//! 8. **Closing the State Channel:** When the allocation period ends, the receiver
//! can send the last RAV to the blockchain and receive payment from the EscrowAccount.
//!
//! ## Performance Considerations
//!
//! - The primary performance limitations are the time required to verify receipts
//! and network limitations for sending requests to the tap_aggregator.
//!
//! ## Use Cases
//!
//! - The TAP protocol is suitable for systems with sequential operations that
//! are too expensive to redeem individually on-chain. By aggregating operations
//! and redeeming them in one transaction, costs can be reduced.
//!
//! ## Compatibility
//!
//! - The current implementation is for EVM-compatible blockchains, with most
//! of the system being off-chain.
//!
//! ## Getting started
//!
//! To get started with the TAP protocol, take a look on the [`manager`] module
//! to see how to manage the state channel and implement the needed adapters.

use std::time::{SystemTime, UNIX_EPOCH};

use alloy_sol_types::eip712_domain;
use thiserror::Error;

mod error;
pub mod manager;
pub mod rav;
pub mod receipt;
pub mod signed_message;

pub use error::Error;
use error::Result;

fn get_current_timestamp_u64_ns() -> Result<u64> {
    Ok(SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map_err(|err| Error::InvalidSystemTime {
            source_error_message: err.to_string(),
        })?
        .as_nanos() as u64)
}

/// The EIP712 domain separator builder for the TAP protocol.
///
/// This is the current domain separator that is used for the [EIP712](https://eips.ethereum.org/EIPS/eip-712) signature scheme.
///
///
/// It's used to validate the signature of the `ReceiptAggregateVoucher` and `Receipt` structs.
///
/// You can take a look on deployed [TAPVerfiers](https://github.com/semiotic-ai/timeline-aggregation-protocol-contracts/blob/4dc87fc616680c924b99dbaf285bdd449c777261/src/TAPVerifier.sol)
/// contracts [here](https://github.com/semiotic-ai/timeline-aggregation-protocol-contracts/blob/4dc87fc616680c924b99dbaf285bdd449c777261/addresses.json)
///
/// The domain separator is defined as:
/// - `name`: "TAP"
/// - `version`: "1"
/// - `chain_id`: The chain ID of the chain where the domain separator is deployed.
/// - `verifying_contract`: The address of the contract that is verifying the signature.
pub fn tap_eip712_domain(
    chain_id: u64,
    verifying_contract_address: alloy_primitives::Address,
) -> alloy_sol_types::Eip712Domain {
    eip712_domain! {
        name: "TAP",
        version: "1",
        chain_id: chain_id,
        verifying_contract: verifying_contract_address,
    }
}

#[cfg(test)]
mod tap_tests {
    use std::str::FromStr;

    use alloy_primitives::Address;
    use alloy_sol_types::Eip712Domain;
    use ethers::signers::{coins_bip39::English, LocalWallet, MnemonicBuilder, Signer};
    use rstest::*;

    use crate::{
        rav::ReceiptAggregateVoucher, receipt::Receipt, signed_message::EIP712SignedMessage,
        tap_eip712_domain,
    };

    #[fixture]
    fn keys() -> (LocalWallet, Address) {
        let wallet: LocalWallet = MnemonicBuilder::<English>::default()
         .phrase("abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about")
         .build()
         .unwrap();
        // Alloy library does not have feature parity with ethers library (yet) This workaround is needed to get the address
        // to convert to an alloy Address. This will not be needed when the alloy library has wallet support.
        let address: [u8; 20] = wallet.address().into();

        (wallet, address.into())
    }

    #[fixture]
    fn allocation_ids() -> Vec<Address> {
        vec![
            Address::from_str("0xabababababababababababababababababababab").unwrap(),
            Address::from_str("0xdeaddeaddeaddeaddeaddeaddeaddeaddeaddead").unwrap(),
            Address::from_str("0xbeefbeefbeefbeefbeefbeefbeefbeefbeefbeef").unwrap(),
            Address::from_str("0x1234567890abcdef1234567890abcdef12345678").unwrap(),
        ]
    }

    #[fixture]
    fn domain_separator() -> Eip712Domain {
        tap_eip712_domain(1, Address::from([0x11u8; 20]))
    }

    #[rstest]
    #[case::basic_rav_test (vec![45,56,34,23])]
    #[case::rav_from_zero_valued_receipts (vec![0,0,0,0])]
    #[test]
    fn signed_rav_is_valid_with_no_previous_rav(
        keys: (LocalWallet, Address),
        allocation_ids: Vec<Address>,
        domain_separator: Eip712Domain,
        #[case] values: Vec<u128>,
    ) {
        // Create receipts
        let mut receipts = Vec::new();
        for value in values {
            receipts.push(
                EIP712SignedMessage::new(
                    &domain_separator,
                    Receipt::new(allocation_ids[0], value).unwrap(),
                    &keys.0,
                )
                .unwrap(),
            );
        }

        // Skipping receipts validation in this test, aggregate_receipts assumes receipts are valid.

        let rav = ReceiptAggregateVoucher::aggregate_receipts(allocation_ids[0], &receipts, None)
            .unwrap();
        let signed_rav = EIP712SignedMessage::new(&domain_separator, rav, &keys.0).unwrap();
        assert!(signed_rav.recover_signer(&domain_separator).unwrap() == keys.1);
    }

    #[rstest]
    #[case::basic_rav_test(vec![45,56,34,23])]
    #[case::rav_from_zero_valued_receipts(vec![0,0,0,0])]
    #[test]
    fn signed_rav_is_valid_with_previous_rav(
        keys: (LocalWallet, Address),
        allocation_ids: Vec<Address>,
        domain_separator: Eip712Domain,
        #[case] values: Vec<u128>,
    ) {
        // Create receipts
        let mut receipts = Vec::new();
        for value in values {
            receipts.push(
                EIP712SignedMessage::new(
                    &domain_separator,
                    Receipt::new(allocation_ids[0], value).unwrap(),
                    &keys.0,
                )
                .unwrap(),
            );
        }

        // Create previous RAV from first half of receipts
        let prev_rav = ReceiptAggregateVoucher::aggregate_receipts(
            allocation_ids[0],
            &receipts[0..receipts.len() / 2],
            None,
        )
        .unwrap();
        let signed_prev_rav =
            EIP712SignedMessage::new(&domain_separator, prev_rav, &keys.0).unwrap();

        // Create new RAV from last half of receipts and prev_rav
        let rav = ReceiptAggregateVoucher::aggregate_receipts(
            allocation_ids[0],
            &receipts[receipts.len() / 2..receipts.len()],
            Some(signed_prev_rav),
        )
        .unwrap();
        let signed_rav = EIP712SignedMessage::new(&domain_separator, rav, &keys.0).unwrap();

        assert!(signed_rav.recover_signer(&domain_separator).unwrap() == keys.1);
    }

    #[rstest]
    #[test]
    fn verify_signature(
        keys: (LocalWallet, Address),
        allocation_ids: Vec<Address>,
        domain_separator: Eip712Domain,
    ) {
        let signed_message = EIP712SignedMessage::new(
            &domain_separator,
            Receipt::new(allocation_ids[0], 42).unwrap(),
            &keys.0,
        )
        .unwrap();

        assert!(signed_message.verify(&domain_separator, keys.1).is_ok());
        assert!(signed_message
            .verify(
                &domain_separator,
                Address::from_str("0x76f4eeD9fE41262669D0250b2A97db79712aD855").unwrap()
            )
            .is_err());
    }
}
