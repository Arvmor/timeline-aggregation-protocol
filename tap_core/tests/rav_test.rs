// Copyright 2023-, Semiotic AI, Inc.
// SPDX-License-Identifier: Apache-2.0

use std::collections::HashMap;
use std::sync::RwLock;
use std::{str::FromStr, sync::Arc};

use alloy_primitives::Address;
use alloy_sol_types::Eip712Domain;
use ethers::signers::coins_bip39::English;
use ethers::signers::{LocalWallet, MnemonicBuilder};
use rstest::*;

use tap_core::manager::context::memory::InMemoryContext;
use tap_core::{
    manager::strategy::{RAVRead, RAVStore},
    rav::ReceiptAggregateVoucher,
    receipt::{checks::TimestampCheck, Receipt},
    signed_message::EIP712SignedMessage,
    tap_eip712_domain,
};

#[fixture]
fn domain_separator() -> Eip712Domain {
    tap_eip712_domain(1, Address::from([0x11u8; 20]))
}

#[fixture]
fn in_memory_context() -> InMemoryContext {
    let escrow_storage = Arc::new(RwLock::new(HashMap::new()));
    let rav_storage = Arc::new(RwLock::new(None));
    let receipt_storage = Arc::new(RwLock::new(HashMap::new()));

    let timestamp_check = Arc::new(TimestampCheck::new(0));
    InMemoryContext::new(
        rav_storage,
        receipt_storage.clone(),
        escrow_storage.clone(),
        timestamp_check,
    )
}

#[rstest]
#[tokio::test]
async fn rav_storage_strategy_test(domain_separator: Eip712Domain, in_memory_context: InMemoryContext) {
    let wallet: LocalWallet = MnemonicBuilder::<English>::default()
         .phrase("abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about")
         .build()
         .unwrap();

    let allocation_id: [u8; 20] = Address::from_str("0xabababababababababababababababababababab")
        .unwrap()
        .into();
    let allocation_id = allocation_id.into();

    // Create receipts
    let mut receipts = Vec::new();
    for value in 50..60 {
        receipts.push(
            EIP712SignedMessage::new(
                &domain_separator,
                Receipt::new(allocation_id, value).unwrap(),
                &wallet,
            )
            .unwrap(),
        );
    }

    let signed_rav = EIP712SignedMessage::new(
        &domain_separator,
        ReceiptAggregateVoucher::aggregate_receipts(allocation_id, &receipts, None).unwrap(),
        &wallet,
    )
    .unwrap();

    in_memory_context.update_last_rav(signed_rav.clone()).await.unwrap();

    // Retreive rav
    let retrieved_rav = in_memory_context.last_rav().await;
    assert!(retrieved_rav.unwrap().unwrap() == signed_rav);

    // Testing the last rav update...

    // Create more receipts
    let mut receipts = Vec::new();
    for value in 60..70 {
        receipts.push(
            EIP712SignedMessage::new(
                &domain_separator,
                Receipt::new(allocation_id, value).unwrap(),
                &wallet,
            )
            .unwrap(),
        );
    }

    let signed_rav = EIP712SignedMessage::new(
        &domain_separator,
        ReceiptAggregateVoucher::aggregate_receipts(allocation_id, &receipts, None).unwrap(),
        &wallet,
    )
    .unwrap();

    // Update the last rav
    in_memory_context.update_last_rav(signed_rav.clone()).await.unwrap();

    // Retreive rav
    let retrieved_rav = in_memory_context.last_rav().await;
    assert!(retrieved_rav.unwrap().unwrap() == signed_rav);
}
