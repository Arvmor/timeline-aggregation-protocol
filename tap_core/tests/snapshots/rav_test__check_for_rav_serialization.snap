---
source: tap_core/tests/rav_test.rs
expression: receipts
snapshot_kind: text
---
[
  {
    "message": {
      "allocation_id": "0xabababababababababababababababababababab",
      "timestamp_ns": 50,
      "nonce": 50,
      "value": 50
    },
    "signature": {
      "r": "0x5257bab234e33525cd999db4defc805c2d3b4e51cde3697f43e37ce39473720f",
      "s": "0x6c3af14c3d400dfd047fd2da90eb9e8cee863e77cc52742ebcbf080b8d6ec2",
      "yParity": "0x1",
      "v": "0x1"
    }
  },
  {
    "message": {
      "allocation_id": "0xabababababababababababababababababababab",
      "timestamp_ns": 51,
      "nonce": 51,
      "value": 51
    },
    "signature": {
      "r": "0x1596dd0d380ede7aa5dec5ed09ea7d1fa8e4bc8dfdb43a4e965bb4f16906e321",
      "s": "0x788b69625a031fbd2e769928b63505387df16e7c51f19ff67c782bfec101a387",
      "yParity": "0x0",
      "v": "0x0"
    }
  },
  {
    "message": {
      "allocation_id": "0xabababababababababababababababababababab",
      "timestamp_ns": 52,
      "nonce": 52,
      "value": 52
    },
    "signature": {
      "r": "0xb3b8e2c1249fc14183024e28b14f7749ef852c898906c2442f380a26bf07a625",
      "s": "0x6925e7dce01d539a658d552e43cfd92d9d204d6997604f8f613977251b964db3",
      "yParity": "0x1",
      "v": "0x1"
    }
  },
  {
    "message": {
      "allocation_id": "0xabababababababababababababababababababab",
      "timestamp_ns": 53,
      "nonce": 53,
      "value": 53
    },
    "signature": {
      "r": "0x3b4d08db319497c2cc0d515d25057e28ce44b194a23e84b9d35682f97027c7e3",
      "s": "0x232e02eb4b52d302d620867a4c10829e5a307404ea1bcbbd2ee33e8422a18a16",
      "yParity": "0x1",
      "v": "0x1"
    }
  },
  {
    "message": {
      "allocation_id": "0xabababababababababababababababababababab",
      "timestamp_ns": 54,
      "nonce": 54,
      "value": 54
    },
    "signature": {
      "r": "0x619d84f659ea3941cdb0656100b2ea8a3d2f5658dbd67f796ebfb8840156530b",
      "s": "0x163b236f88207b89452255da8ce196997d8f2f0880081659c780f2093797f75e",
      "yParity": "0x1",
      "v": "0x1"
    }
  },
  {
    "message": {
      "allocation_id": "0xabababababababababababababababababababab",
      "timestamp_ns": 55,
      "nonce": 55,
      "value": 55
    },
    "signature": {
      "r": "0x48e1e0e31eaf40eabbcbc3c6d125b7656c0796d51188f89d27194e22f2c5d6bb",
      "s": "0xd26efc0ae8cc3646993a20b5aabac1125ecb149ad91d733c702ac9f03222b66",
      "yParity": "0x1",
      "v": "0x1"
    }
  },
  {
    "message": {
      "allocation_id": "0xabababababababababababababababababababab",
      "timestamp_ns": 56,
      "nonce": 56,
      "value": 56
    },
    "signature": {
      "r": "0xc3adb8be5db130f563d3a18fc9e742fca84f69a903413e04dc567b9c3aca8626",
      "s": "0x564dd73bdd33897c7a085e4eb1bc0ce002b1d65c6006781ab54cd670846fe358",
      "yParity": "0x0",
      "v": "0x0"
    }
  },
  {
    "message": {
      "allocation_id": "0xabababababababababababababababababababab",
      "timestamp_ns": 57,
      "nonce": 57,
      "value": 57
    },
    "signature": {
      "r": "0xa0fe51e1b7253daed14f99c9320d0a539ef18b6ead6552947e5c93dde6f40dea",
      "s": "0x4277d66d3a8c9f67cddc8d96a71ef8437e47e34a5b5789d7843eb691c3b9864",
      "yParity": "0x0",
      "v": "0x0"
    }
  },
  {
    "message": {
      "allocation_id": "0xabababababababababababababababababababab",
      "timestamp_ns": 58,
      "nonce": 58,
      "value": 58
    },
    "signature": {
      "r": "0x26f1657e4b8759867820be12c25e982e15ff9d70aa99fe1fd2587cbb644829de",
      "s": "0x5cabbd965f93e544b07e5956c2831148dbf7960b4e2edadfa6ecbf1209dacda4",
      "yParity": "0x0",
      "v": "0x0"
    }
  },
  {
    "message": {
      "allocation_id": "0xabababababababababababababababababababab",
      "timestamp_ns": 59,
      "nonce": 59,
      "value": 59
    },
    "signature": {
      "r": "0x90ce08049b9ce9fa38077ebeed0e24558442d8ae001aeff6b9f4b06f4f553c69",
      "s": "0x7a873491448ae696555f9d1314b4e78a7cc98a19a6b0c26aad562053dc26a202",
      "yParity": "0x1",
      "v": "0x1"
    }
  }
]
